### ISSUE TITLE
Boot hangs due to non-architected eieio store forwarding barrier on older Qemus

### ISSUE THREAD

**Issue by mpe** [linuxppc/issues#160]
Using non-mainline qemu, the boot hangs when using CPU POWER9.

eg. the version in Ubuntu 17.10:
```
$ qemu-system-ppc64 --version
QEMU emulator version 2.10.1(Debian 1:2.10+dfsg-0ubuntu3.8)
Copyright (c) 2003-2017 Fabrice Bellard and the QEMU Project developers
$ qemu-system-ppc64 -nographic -vga none -M pseries -cpu POWER9 -m 1G -kernel build/vmlinux -initrd $HOME/work/topics/root-disks/ppc64le-rootfs.cpio.gz 
...
bootconsole [udbg0] disabled
bootconsole [udbg0] disabled
pid_max: default: 32768 minimum: 301
Dentry cache hash table entries: 131072 (order: 4, 1048576 bytes)
Inode-cache hash table entries: 65536 (order: 3, 524288 bytes)

<hangs>

QEMU 2.10.1 monitor - type 'help' for more information
(qemu) info registers
NIP c000000000004714   LR c0000000000cb11c CTR c00000000066b6e0 XER 0000000020040000 CPU#0
MSR 8000000000001031 HID0 0000000000000000  HF 8000000000000031 iidx 1 didx 1
TB 00000003 14743762458 DECR 2286072171
GPR00 c00000000066b72c c00000000123fa60 c000000001242700 0000000000000058
GPR04 0000000071000000 0000000000000010 4d6f756e742d6361 6368652068617368
GPR08 c00000000123faff 0010000000000000 c0000000000cb118 0000000000000000
GPR12 c00000000066b6e0 c0000000013d0000 000000003dc5bef0 0000000000000000
GPR16 000000000000003d 20c49ba5e353f7cf 0000000000000000 c0000000012d98b8
GPR20 c0000000012d78b8 0000000000000000 000000000000000d 0000000000000000
GPR24 0000000000000010 c0000000012d74b0 0000000000000000 c0000000011b2168
GPR28 000000000000003d 0000000000000000 000000000000002d 0000000000000010
CR 28022222  [ E  L  -  E  E  E  E  E  ]             RES ffffffffffffffff
FPR00 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR04 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR08 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR12 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR16 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR20 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR24 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR28 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPSCR 0000000000000000
 SRR0 c000000000004724  SRR1 8000000000081031    PVR 00000000004e0000 VRSAVE 0000000000000000
SPRG0 0000000000000000 SPRG1 c0000000013d0000  SPRG2 c0000000013d0000  SPRG3 0000000000000000
SPRG4 0000000000000000 SPRG5 0000000000000000  SPRG6 0000000000000000  SPRG7 0000000000000000
HSRR0 0000000000000000 HSRR1 0000000000000000
 CFAR c0000000000cb118
 LPCR 0000000007c1f40c
  DAR 0000000000000000  DSISR 0000000000000000
```

That shows us in program check, and the CFAR is pointing into `hvc_put_chars()`.

Hopefully this is just something missing/wrong in that version of qemu, but it would still be good to diagnose exactly what and document it.

**Comment by mpe** [comment]
In summary, the kernel change to add the store forwarding barrier using `eieio` (torvalds/linux@a048a07d7f4535baa4cbad6bc024f175317ab938) broke booting on Qemu because it used a previously undefined and still undocumented form of the `eieio` instruction. That was fixed in qemu commit [c8fd8373e42821984400382cd91b8bf4e7c14e3b](https://git.qemu.org/?p=qemu.git;a=commitdiff;h=c8fd8373e42821984400382cd91b8bf4e7c14e3b).

**Comment by mpe** [comment]
Thanks to @mopsfelder for chasing it up! :cake:
